core/src/main/java/com/linecorp/armeria/client/Bootstraps.java

```diff
@@ -0,0 +1,136 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static com.linecorp.armeria.common.SessionProtocol.httpAndHttpsValues;
+
+import java.lang.reflect.Array;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.Set;
+
+import com.linecorp.armeria.common.SerializationFormat;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoop;
+import io.netty.channel.unix.DomainSocketAddress;
+import io.netty.handler.ssl.SslContext;
+
+final class Bootstraps {
+    private final Bootstrap[][] inetBootstraps;
+    @Nullable
+    private final Bootstrap[][] unixBootstraps;
+
+    private final int connectTimeoutMillis;
+
+    private final EventLoop eventLoop;
+
+    private final SslContext sslCtxHttp1Only;
+
+    private final SslContext sslCtxHttp1Or2;
+
+    Bootstraps(HttpClientFactory clientFactory, EventLoop eventLoop, SslContext sslCtxHttp1Only,
+               SslContext sslCtxHttp1Or2) {
+        this.eventLoop = eventLoop;
+        this.sslCtxHttp1Only = sslCtxHttp1Only;
+        this.sslCtxHttp1Or2 = sslCtxHttp1Or2;
+        final Bootstrap inetBaseBootstrap = clientFactory.newInetBootstrap();
+        final Bootstrap unixBaseBootstrap = clientFactory.newUnixBootstrap();
+        inetBootstraps = newBootstrapMap(inetBaseBootstrap, clientFactory, eventLoop);
+        if (unixBaseBootstrap != null) {
+            unixBootstraps = newBootstrapMap(unixBaseBootstrap, clientFactory, eventLoop);
+        } else {
+            unixBootstraps = null;
+        }
+        connectTimeoutMillis = (Integer) inetBaseBootstrap.config().options()
+                                                          .get(ChannelOption.CONNECT_TIMEOUT_MILLIS);
+    }
+
+    Bootstrap get(SocketAddress remoteAddress, SessionProtocol desiredProtocol,
+                  SerializationFormat serializationFormat) {
+        if (!httpAndHttpsValues().contains(desiredProtocol)) {
+            throw new NullPointerException("Unsupported session protocol: " + desiredProtocol);
+        }
+
+        if (remoteAddress instanceof InetSocketAddress) {
+            return select(inetBootstraps, desiredProtocol, serializationFormat);
+        }
+
+        assert remoteAddress instanceof DomainSocketAddress : remoteAddress;
+
+        if (unixBootstraps == null) {
+            throw new IllegalArgumentException("Domain sockets are not supported by " +
+                                               eventLoop.getClass().getName());
+        }
+
+        return select(unixBootstraps, desiredProtocol, serializationFormat);
+    }
+
+    int getConnectTimeoutMillis() {
```

Comment:
The `connectTimeoutMills` came from `inetBaseBootstrap.config`, so I added that value to `Bootstraps`. If I wanted to get that value from the `HttpChannelPool`, I would have to create a new `inetBootstraps` in `cleintFactory`, so that's how I did it. But conceptually, having it in the `httpChannelPool` looks good!!!  😄

Reply:
What do you think about taking `clientFacotry`, `inetBootstraps`, and `unixBootstraps` as arguments to the constructor of `Bootstraps`?

This would solve the problem I was having with having to create duplicates.

Reply:
How about using `clientFactory` options so that we don't have to create `inetBaseBootstrap` in `HttpChannelPool`?
```java
HttpChannelPool(HttpClientFactory clientFactory, EventLoop eventLoop,
                SslContext sslCtxHttp1Or2, SslContext sslCtxHttp1Only,
                ConnectionPoolListener listener) {
    ...
    this.listener = listener;

    connectTimeoutMillis =
            (Integer) clientFactory.options().channelOptions().get(ChannelOption.CONNECT_TIMEOUT_MILLIS);

    bootstraps = new Bootstraps(clientFactory, eventLoop, sslCtxHttp1Or2, sslCtxHttp1Only);
}

Bootstraps(HttpClientFactory clientFactory, EventLoop eventLoop, SslContext sslCtxHttp1Or2,
           SslContext sslCtxHttp1Only) {...}
```

Reply:
This isn't resolved. 😉

Reply:
I missed it. Thanks @minwoox. I didn't know that I could get `connectTimeoutMills` by `clientFactor.options()` . I'll apply that and post back.